### ISSUE TITLE
GRPO finetune with LORA leads to poor performance

### ISSUE THREAD

**Issue by zaddy6** [huggingface/trl#2872]
### Reproduction

![Image](https://github.com/user-attachments/assets/95f3abad-b2c2-48b5-9490-28290a5fc62a)

I noticed training without LORA leads to better performance, here is an example without LORA it starts to max the rewards at 1k steps, with Lora it doesnt learn

Model is Qwen2.5-3B

```Current config

lora_config = LoraConfig(
r=8,
lora_alpha=16,
target_modules="all-linear",
lora_dropout=0.05,
use_dora=True,
)


training_args = GRPOConfig(
    output_dir=output_dir,
    run_name=run_name,
    learning_rate=5e-7,
    lr_scheduler_type='cosine',
    logging_steps=1,
    per_device_train_batch_size=2,
    gradient_accumulation_steps=1,
    gradient_checkpointing=True,
    gradient_checkpointing_kwargs={"use_reentrant": False},
    bf16=True,
    num_generations=2,
    max_prompt_length=512,
    max_completion_length=512,
    use_vllm = True,
    num_train_epochs=4,
    save_steps=100,
    report_to="wandb",
    save_total_limit=2,
    log_on_each_node=False,
    beta=0.001,
    warmup_ratio=0.1,
    vllm_gpu_memory_utilization=0.7,
    optim="adamw_8bit",
    # temperature=0.5
)


```

### System Info

- Platform: Linux-5.15.0-130-generic-x86_64-with-glibc2.35
- Python version: 3.11.9
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3, NVIDIA H100 80GB HBM3
- Transformers version: 4.48.2
- Accelerate version: 1.3.0
- Accelerate config: not found
- Datasets version: 3.2.0
- HF Hub version: 0.28.1
- TRL version: 0.16.0.dev0
- bitsandbytes version: 0.45.2
- DeepSpeed version: not installed
- Diffusers version: not installed
- Liger-Kernel version: 0.5.2
- LLM-Blender version: not installed
- OpenAI version: 1.61.1
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by af-74413592** [comment]
code totally wrong

**Comment by af-74413592** [comment]
> code totally wrong

this is not GRPO.

**Comment by af-74413592** [comment]
> > code totally wrong
> 
> this is not GRPO.

just a policy gradient

**Comment by AIR-hl** [comment]
@zaddy6 peft+vllm may not work correctly. #2856

**Comment by af-74413592** [comment]
![Image](https://github.com/user-attachments/assets/52341bcb-8317-4ef1-ad3c-04de3d74728f)

**Comment by FormoJ** [comment]
whats your reward model

**Comment by Maghoumi** [comment]
See this one for a solution: https://github.com/huggingface/trl/issues/2856#issuecomment-2661150151

I confirm with that fix, LoRA + GRPO works well

**Comment by auto-Dog** [comment]
> > > code totally wrong
> > 
> > 
> > this is not GRPO.
> 
> just a policy gradient

What's your point? you mean num_generation=2 (which makes reward -1 and 1 per group) break GRPO hypothesis?

**Comment by zjc664656505** [comment]
Same issue here. My model has lower performance using Lora and vllm after training. Not sure why this is happening being honest. My num_generation is set to 8. I think based on previous threads, it's better to disable the lora config and train the model with standard sft and do not use the vllm. I would give a try again to see the performance. Any thoughts?

**Comment by purbeshmitra** [comment]
Are you able to solve this issue? Any fix? @zaddy6 @zjc664656505

**Comment by qgallouedec** [comment]
Hey @zaddy6, sorry for the very slow reply, we're catching up on the backlog.

This is the known LoRA + vLLM weight-sync bug tracked in #2856: vLLM was being loaded with the *base* weights because the adapter got unmerged before `load_weights()` ran, so generation effectively ignored your LoRA. Which is exactly why "with LoRA it doesn't learn."

It was fixed in #2873, merged a couple of days after you filed this, so your `0.16.0.dev0` snapshot landed right in the gap and didn't have it. On current main the sync correctly does merge → load → unmerge, see https://github.com/huggingface/trl/blob/bd128c5e4dd6b1f78c0623b75263630a3ebf61f6/trl/generation/vllm_generation.py#L454-L480 and LoRA + vLLM GRPO trains as expected.

Could you upgrade to the latest release and retry? Two unrelated tuning notes: `num_generations=2` gives a very weak advantage signal (each group collapses to roughly ±1), and `5e-7` is low for LoRA; both worth bumping.
